### ISSUE TITLE
Expose counters for torch.distributed to 

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#43238]
## 🚀 Feature
Similar to `torch.cuda.memory_stats()`, `torch.distributed` should provide functionality in order to report the total communication overhead within a scope, e.g. within a model's forward/backward pass. For example, a counter that reports the total inter-rank communication latency within a block and provides a way to reset that counter would be quite useful.

## Motivation
This feature request comes @vreis. This will allow performance profiling and debugging such as evaluating the comm. overhead of different algorithms, and gauging the amount of overlap between comm. and backwards computation, for example in DDP's backwards pass. 

## Pitch

Since this should probably not be "always-on" due to the possible extra overhead, a context manager such as `with torch.distributed.collect_comm_stats() as comm_stats` could be useful. We could then aggregate the comm. time for any `torch.distributed` collectives called under this context manager. Something like `comm_stats.get_total_comm_overhead()` could return the total communication overhead within the block. 

In order to make this work as expected, with the above proposal the user would most likely have to `wait()` on all async work, such as allreduce, that was created to finish (and synchronize in the case of nccl operations). If this requires too much change to user code to be convenient to use, we may want to eventually provide a `block_until_outsanding_work_complete` or similar flag in the context manager, where we will record all work created and wait on it when exiting (this is similar to the `rpc.wait_all()`: https://github.com/pytorch/pytorch/pull/41807

**Comment by pritamdamania87** [comment]
I'm wondering if addressing #43231 is sufficient for the use case here? Or would we like to avoid enabling the autograd profiler and only collect some stats about communication overhead without adding a lot of overhead from the autograd profiler.

**Comment by rohan-varma** [comment]
> I'm wondering if addressing #43231 is sufficient for the use case here? Or would we like to avoid enabling the autograd profiler and only collect some stats about communication overhead without adding a lot of overhead from the autograd profiler.

From a comment by @vreis  (internal) it was mentioned that the profiler is not always on, I’m not sure if this is only to avoid overhead or another reason. But due to that, thought it was worth creating another issue for this specifically. The benefit as you mentioned is we remove the overhead of recording each op that the profiler would, and it would probably be a bit easier on the user to only have the comm. stats if that's what they're interested in (at the cost of having to learn one more api, however).
